src/ui/ui-actions.js

```
@@ -11,18 +11,60 @@ const actionButton = $("action-button");
 const configButton = $("config-button");
 const configPanel = $("config-panel");
 const deviceFrame = $("device-frame");
+const downloadLogsButton = $("download-logs-button");
+
+var logsList = [];
 
 configButton.addEventListener("click", () => {
   configPanel.classList.toggle("visible");
 });
 
+function downloadFile(filename, text) {
+  var element = document.createElement("a");
+  element.setAttribute(
+    "href",
+    "data:text/plain;charset=utf-8," + encodeURIComponent(text),
+  );
+  element.setAttribute("download", filename);
+  element.style.display = "none";
+  document.body.appendChild(element);
+  element.click();
+  document.body.removeChild(element);
+}
+
+downloadLogsButton.addEventListener("click", () => {
+  logsOutputText = ``;
+  errorsCount = 0;
+  logsList.forEach(function(entry) {
+    if (entry.className == "error") {
```

Can we also propagate the errors to the top of the document (as well as leaving them in place) just so the top of the document has all the errors at a glance?